pyccel/codegen/printing/fcode.py

```diff
@@ -2608,31 +2608,77 @@ def _print_IndexedElement(self, expr):
+                inds[i] = self._slice_management(ind, _shape, allow_negative_indexes)
+            elif isinstance(ind, PyccelUnarySub) and isinstance(ind.args[0], LiteralInteger):
+                inds[i] = PyccelMinus(_shape, ind.args[0])
             else:
                 #indices of indexedElement of len==1 shouldn't be a Tuple
                 if isinstance(ind, Tuple) and len(ind) == 1:
                     inds[i] = ind[0]
                 if allow_negative_indexes and not isinstance(ind, LiteralInteger):
-                    inds[i] = PyccelMod(ind, base_shape[i])
+                    inds[i] = PyccelMod(ind, _shape)
 
         inds = [self._print(i) for i in inds]
 
-        return "%s(%s)" % (base, ", ".join(inds))
+        return "%s(%s)" % (base_code, ", ".join(inds))
 
 
     def _print_Idx(self, expr):
         return self._print(expr.label)
 
+    def _slice_management(self, _slice, shape, allow_negative_index):
+        start = _slice.start
+        end = _slice.end
+        step = _slice.step
+
+        # negative start and end in slice
+        if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
+            start = PyccelMinus(shape, start.args[0])
+        elif start is not None and allow_negative_index and not isinstance(start,LiteralInteger):
+            start = IfTernaryOperator(PyccelLt(start, LiteralInteger(0)),
+                        PyccelAdd(shape, start), start)
+
+        if isinstance(end, PyccelUnarySub) and isinstance(end.args[0], LiteralInteger):
+            end = PyccelMinus(shape, end.args[0])
+        elif end is not None and allow_negative_index and not isinstance(end, LiteralInteger):
+            end = IfTernaryOperator(PyccelLt(end, LiteralInteger(0)),
+                        PyccelAdd(shape, end), end)
+
+        tmp_end = end #save temporary value of end
+        # steps in slices
+        if step is not None :
+            #negative step in slice
+            if isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
+                end = PyccelAdd(end, LiteralInteger(1)) if end is not None else LiteralInteger(0)
+                start = start if start is not None else PyccelMinus(shape, LiteralInteger(1))
+
+            # variable step in slice
+            elif allow_negative_index and not isinstance(step, LiteralInteger):
+                if start is None :
+                    start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)),
+                        LiteralInteger(0), PyccelMinus(shape, LiteralInteger(1)))
+
+                if end is None :
+                    end = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)),
+                        PyccelMinus(shape, LiteralInteger(1)), LiteralInteger(0))
+                else :
+                    end = PyccelAdd(end, LiteralInteger(1))
+
+        if end is not None and end == tmp_end:
+            end = PyccelMinus(end, LiteralInteger(1))
```

Comment:
This doesn't seem like the clearest way of doing this to me. Can you explain what is happening and why you need this `if` block?
Do you not get: `end+1-1` in the end? for variable steps in slice?
Can you not do:
`if end is not None:`
before the block above and then just not touch cases that are already correct?

Reply:
I have done this just to avoid writing end + 1 - 1

Reply:
Is there no way to make this an `else` instead of checking if `end == tmp_end` to see if it is modified?

Reply:
Ah, so it needs to be done in this order until issue #564 is fixed?

Reply:
Is there no way to make this an else instead of checking if end == tmp_end to see if it is modified?

Reply:
Eg.
```
if isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
    # negative step in slice
    ...
elif step is None or not allow_negative_index:
    # positive step in slice
    stop = PyccelMinus(stop, LiteralInteger(1))
else:
    # general step in slice
    ...
```
?

Reply:
yes, changing order might work.